[ferroamp] Binding for ferroamp

basse04 opened the pull request:
This is the binding for ferroamp, 20240109

This ferroamp binding will connect to a ferroamp Energy System and collect information regarded to the Energy Hub. It is also possible to set some parameters, all described in README.

Br Basse

basse04 commented:
Hi, I think I found the reason for CI / Build.....Failing after....
Found this eminent help written by hilbrand:
"Because you have missing entries in CODEOWNERS and bom/openhab-addons/pom.xml"
I'm using Eclipse IDE, and my question is now how I will edit and Stage these files.
I found in them in my repo and branch as ferroamp_v3, do you think I will import them to Eclipse / Edit / Commit and Stage as usual, or do I have to use some other method?
Br Basse

lsiepel commented:
> found in them in my repo and branch as ferroamp_v3, do you think I will import them to Eclipse / Edit / Commit and Stage as usual, or do I have to use some other method?

Just edit/save/commit as any other file.

basse04 commented:
Hi,
again and again, Thank you so very much for all help. This is not what I as a no-pro is doing  every day.
And now it seems like all checks have passed, so far so good!

Br Basse

lsiepel commented on bundles/org.openhab.binding.ferroamp/README.md:
```diff
@@ -0,0 +1,203 @@
+Discovery is not supported.
+
+## Thing Configuration
+
+Installing of binding ferroamp and configuration of it can be done in the UI.
```
```suggestion
```

lsiepel replied:
Still there.

lsiepel commented on bundles/org.openhab.binding.ferroamp/README.md:
```diff
@@ -0,0 +1,203 @@
+
+Installing of binding ferroamp and configuration of it can be done in the UI.
+
+The following configuration-parameters are available for the `ferroamp` binding.
```
```suggestion
The following configuration parameters are available:
```

lsiepel commented on bundles/org.openhab.binding.ferroamp/README.md:
```diff
@@ -0,0 +1,203 @@
+| hasBattery      | boolean | Is there a battery connected or not ? | N/A     | no       | yes      |
+
+## Channels
+
+The following channels are available for reading, for the `ferroamp` binding.
```
```suggestion
```

lsiepel commented on bundles/org.openhab.binding.ferroamp/README.md:
```diff
@@ -0,0 +1,203 @@
+| esm-status           | String                     | R          | Dependent on battery manufacturer        |
+| esm-ts               | String                     | R          | Time stamp when message was published    |
+
+The following channels are available for `ferroamp` energy system configuration. Please, see Ferroamp documentation for more details.
```
These channels are just like any other, they are just write channels. Maybe add them to the table aboce? Also are they write only or R/W ?

lsiepel commented on bundles/org.openhab.binding.ferroamp/README.md:
```diff
@@ -0,0 +1,203 @@
+
+## Thing
+
+Thing ferroamp:energyhub:myenergyhub [ hostName="'energyhub-ip'", userName="'myUserName'", password="'myPassword'", hasBattery=false ]
```
```suggestion
```java
Thing ferroamp:energyhub:myenergyhub [ hostName="'energyhub-ip'", userName="'myUserName'", password="'myPassword'", hasBattery=false ]
\```
```

lsiepel replied:
```suggestion
Thing ferroamp:energyhub:myenergyhub [hostName="energyhub-ip", userName="myUserName", password="myPassword", hasBattery=false]
```
Minor, but for clarity, please remove the inner single quotes

lsiepel commented on bundles/org.openhab.binding.ferroamp/README.md:
```diff
@@ -0,0 +1,203 @@
+
+## Items
+
+Number:Energy Ferroamp "ehub-wloadconsq-l1" <energy> { channel="ferroamp:energyhub:myenergyhub:ehub-wloadconsq-l1" }
+String Ferroamp "request-charge" <energy> { channel="ferroamp:energyhub:myenergyhub:request-charge" }
```
Can you also add this to code fences ?

lsiepel commented on bundles/org.openhab.binding.ferroamp/README.md:
```diff
@@ -0,0 +1,203 @@
+Thing ferroamp:energyhub:myenergyhub [ hostName="'energyhub-ip'", userName="'myUserName'", password="'myPassword'", hasBattery=false ]
+
+## Items
+
```
The readme.md is missing a sitemap example sections, can you add that too?

lsiepel replied:
Still missing

basse04 replied:
Added a part regarded Rules

lsiepel replied:
The example only has two items, but ok.

lsiepel commented on bundles/org.openhab.binding.ferroamp/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,34 @@
+
+	<!-- ferroamp Energyhub Thing Type -->
+	<thing-type id="energyhub">
+		<label>ferroamp Binding Thing</label>
```
```suggestion
		<label>Ferroamp Energy Hub Thing</label>
```

lsiepel replied:
Not fixed.

basse04 replied:
Fixed

lsiepel commented on bundles/org.openhab.binding.ferroamp/src/main/resources/OH-INF/thing/Ferroamp.xml:
```diff
@@ -0,0 +1,161 @@
+		<description>Bindings for the ferroamp Energy Hub</description>
+		<category>EnergyHub</category>
+		<channels>
+			<channel typeId="ehub-wloadconsq-l1" id="ehubWLoadConsqL1"/>
```
Can you adjust all channel id's to lower-case-hyphen
https://www.openhab.org/docs/developer/guidelines.html#naming-convention

basse04 replied:
Please, see commit 20240607

basse04 replied:
Please, see commit 20240607

lsiepel commented on bundles/org.openhab.binding.ferroamp/src/main/resources/OH-INF/thing/Channels.xml:
```diff
@@ -0,0 +1,978 @@
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+	<channel-type id="ehub-wloadconsq-l1">
+		<item-type>Number:Energy</item-type>
+		<label>EHUB, WLoadConsq L1</label>
```
All channels have the label prefix EHUB, this can be removed as it adds no value.

basse04 replied:
Please, see commit 20240607

lsiepel commented on bundles/org.openhab.binding.ferroamp/src/main/resources/OH-INF/thing/Channels.xml:
```diff
@@ -0,0 +1,978 @@
+	</channel-type>
+	<channel-type id="ehub-wloadconsq-l2">
+		<item-type>Number:Energy</item-type>
+		<label>EHUB, WLoadConsq L2</label>
```
What does `WLoadConsq` mean? Is it possible to have give it a more meaningfull and easy to read (short) label?

Please check others too.

basse04 replied:
WLoadConsq and all the others are the name of the parameter I have found in the specifications of the system. I think it will be best to use them. For those who has an description in the spec. are they as: <description> in channels.xml

lsiepel commented on bundles/org.openhab.binding.ferroamp/src/main/resources/OH-INF/i18n/ferroamp.properties:
```diff
@@ -0,0 +1,3 @@
+# FIXME: please add all English translations to this file so the texts can be translated using Crowdin
```
One of the last steps after the review changes are checked in, the i18n plugin should be executed:
https://www.openhab.org/docs/developer/utils/i18n.html#generating-i18n-properties-file

lsiepel commented on bundles/org.openhab.binding.ferroamp/src/main/resources/OH-INF/addon/addon.xml:
```diff
@@ -0,0 +1,10 @@
+	<type>binding</type>
+	<name>ferroamp Binding</name>
+	<description>This is the binding for ferroamp energyhub.</description>
+
```
Can you add the connection tag and if possible the country tag?

lsiepel reviewed (changes requested):
Thanks for your contribution! 
First pass, looked at 15/22 files. Mainly documentation, i18n, thing structure etc., handlers etc will follow later.

I did some suggestions regarding the name, not sure if im right. The binding should be consistent in the use of some names like ferroamp (capitalized or not)  and the use of energy hub vs energyhub (also proper capitalized).

lsiepel commented:
@basse04 ping me when you are ready for another review pass.

basse04 commented:
Hi,
ping :-)
I have just uploaded some more changes.
About I18N so will I be back, seems like I have to do some reading about this.
About Channel-prefix so is it as I can understand necessery to have  the first part there though it is for some channels the very same name but regarded to different parts of system. Ex. ehub-wpv and ssos0-wpv.
About wloadconsq so will I have to do some more checks, I have just follow the parameter-names described in manual, some parameters doesn't have any explanation at all.

There are Checks failed, I will have a look.

BR Basse

basse04 commented:
Hi,
just added ferroamp.properties and done some changes in channels.xml regarded to this.
I still have some thoughts about Label / Description text, so Please have a look and let me know what you think about that.
Br Basse

lsiepel commented on bom/openhab-addons/pom.xml:
```diff
@@ -7,7 +7,7 @@
     <groupId>org.openhab.addons.bom</groupId>
     <artifactId>org.openhab.addons.reactor.bom</artifactId>
-    <version>4.0.0-SNAPSHOT</version>
+    <version>4.2.0-SNAPSHOT</version>
```
Can you revert this change?

jlaur replied:
Sorry to interfere, but it seems this is not yet reverted.

lsiepel commented on bom/pom.xml:
```diff
@@ -7,7 +7,7 @@
     <groupId>org.openhab.addons</groupId>
     <artifactId>org.openhab.addons.reactor</artifactId>
-    <version>4.0.0-SNAPSHOT</version>
+    <version>4.2.0-SNAPSHOT</version>
```
Can you revert changes to this file?

lsiepel reviewed (changes requested):
Can you fix the comments ? Labels are better now. As there are so many channels, it might be posssible to re-use some channel type id's. Please also adjust the naming convention as per earlier review comment.
When the comments are addressed and you are ready for a next review round, let me know.

basse04 commented:
Hi, Isiepel.
Please have a look at my latest  changes regarded to your (as always very good and informative) comments. The changes are done in pom.xml and README.md.
About your thoughts about configuration-channels so are they only W, I have had a look at re-using channel-id's, but I think we have to have unique id's for each due to that they are reading values from different parts of the system and every system can be 
built up of different possible parts as customers needs.

BR Basse

lsiepel commented:
> Hi, Almost forgot to change /bom/pom.xml, now I have done. BR Basse

Unfortunately i don't see any changes. If possible don't use force push. Just commit and push if possible.

basse04 commented:
Hi,
Gh sometimes drives me nuts :-)
I use to Commit from Eclipse and often it works as expected, but sometimes I got the Faultmessage like non-fast forward. I then use the Force Push option and it ususally then works.
I have read about this problem, and it can be regarded to "the remote branch has evolved since you started your development"
but I have not done any changes since last Commit, hm?
Anyway, Please have a look. I can see the changes at https://github.com/openhab/openhab-addons/pull/16258/commits/d99fd22755a8b6f80bcff6bb8ed237731d98852b and https://github.com/openhab/openhab-addons/pull/16258/commits/38718b487f92c0e04839c3f8b23394144b174e48
I really hope that this is were they should be.
BR Basse

lsiepel commented:
> Hi, Gh sometimes drives me nuts :-) I use to Commit from Eclipse and often it works as expected, but sometimes I got the Faultmessage like non-fast forward. I then use the Force Push option and it ususally then works. I have read about this problem, and it can be regarded to "the remote branch has evolved since you started your development" but I have not done any changes since last Commit, hm? Anyway, Please have a look. I can see the changes at [d99fd22](https://github.com/openhab/openhab-addons/commit/d99fd22755a8b6f80bcff6bb8ed237731d98852b) and [38718b4](https://github.com/openhab/openhab-addons/commit/38718b487f92c0e04839c3f8b23394144b174e48) I really hope that this is were they should be. BR Basse

Issues that are resolved are marked, would be nice if we can get this sorted in the next week as we are getting closer to the next stable release.

basse04 commented:
It sounds good so far.
And as always, just let me know if there is anything I can do.

lsiepel commented on bundles/org.openhab.binding.ferroamp/README.md:
```diff
@@ -0,0 +1,203 @@
+Binding MQTT must be installed before binding ferroamp is able to use.
+
+Contact ferroamp support to enable MQTT in the EnergyHub and to get Username and Password:
+https://ferroamp.com/om-ferroamp/
```
Not fixed yet

lsiepel replied:
Still not fully fixed. The `the` is missing before username.

lsiepel commented:
> It sounds good so far. And as always, just let me know if there is anything I can do.

Could you verify and respond to the comments that are not yet marked as resolved? I will then mark them as resolved once i verified they are fine. This is vfor me the best procedure for reviewing.

basse04 commented:
Hi, I must give you my deepest excuses, I have totally missed some of your comments. I have now taking care of them and this changes should be in my latest commit.
About CODEOWNERS so, Plese help me, I don't understand what I have done wrong.
About ferroamp.properties so will I need some more time, maybe this is not very important?

I see a fault message in CI / Build(Java 21.....Seems to be regarded to org.openhab.binding.http ?

Is it OK if I Comment as this or do you want me to Reply on each Commit?

BR Basse

basse04 commented:
Hi, 
I have done some more changes, regarded thing-types.xml, Channels.xml and ferroamp.properties and all checks have passed.
So Please, have a look.
BR Basse

jlaur commented on bundles/org.openhab.binding.ferroamp/src/main/resources/OH-INF/addon/addon.xml:
```diff
@@ -0,0 +1,11 @@
+
+	<type>binding</type>
+	<name>ferroamp Binding</name>
+	<description>This is the binding for ferroamp energyhub.</description>
```
```suggestion
	<description>This is the binding for Ferroamp EnergyHub.</description>
```

(found https://ferroamp.com/en/products/energyhub/)

basse04 replied:
Fixed

jlaur commented on bundles/org.openhab.binding.ferroamp/src/main/resources/OH-INF/addon/addon.xml:
```diff
@@ -0,0 +1,11 @@
+	xsi:schemaLocation="https://openhab.org/schemas/addon/v1.0.0 https://openhab.org/schemas/addon-1.0.0.xsd">
+
+	<type>binding</type>
+	<name>ferroamp Binding</name>
```
```suggestion
	<name>Ferroamp Binding</name>
```

jlaur commented on bundles/org.openhab.binding.ferroamp/src/main/resources/OH-INF/thing/Ferroamp.xml:
```diff
@@ -0,0 +1,161 @@
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+	<thing-type id="energyhub">
+		<label>Ferroamp EnergyHub</label>
+		<description>Bindings for the Ferroamp EnergyHub</description>
```
This is a Thing type description, so "Bindings for" seems wrong here. This is only a suggestion for change:
```suggestion
		<description>Represents the Ferroamp EnergyHub</description>
```

jlaur commented on bundles/org.openhab.binding.ferroamp/src/main/resources/OH-INF/thing/Channels.xml:
```diff
@@ -0,0 +1,978 @@
+		<category>Energy</category>
+		<state readOnly="true" pattern="%.2f %unit%"/>
+	</channel-type>
+	<channel-type id="ehub-wloadconsq-l2">
```
You do not need to define multiple channel types for the same purpose just to have different labels and descriptions. Instead, you can override the label and description in the channels referencing the channel type.

There are also system channel types you could consider using - see:
https://github.com/openhab/openhab-core/blob/main/bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/DefaultSystemChannelTypeProvider.java

basse04 replied:
Done in channels.xml

basse04 replied:
Please, see commit 20240607

jlaur commented on bundles/org.openhab.binding.ferroamp/src/main/resources/OH-INF/thing/Channels.xml:
```diff
@@ -0,0 +1,978 @@
+		<state readOnly="true" pattern="%.2f %unit%"/>
+	</channel-type>
+	<channel-type id="ehub-ts">
+		<item-type>String</item-type>
```
For timestamps, item type `DateTime` would be a better match:
```suggestion
		<item-type>DateTime</item-type>
```

jlaur commented on bundles/org.openhab.binding.ferroamp/src/main/resources/OH-INF/thing/Channels.xml:
```diff
@@ -0,0 +1,978 @@
+		<item-type>String</item-type>
+		<label>EHUB, Time Stamp</label>
+		<description>Time stamp when message was published</description>
+		<category>Energy</category>
```
```suggestion
		<category>Time</category>
```

lsiepel reviewed (changes requested):
Found these comments, please look if they can also apply on other places in the binding.
Ping whenever you are ready for another review pass.

basse04 commented:
Hi,
Please see latest changes in 24189d5.
I will continue with the rest...
BR Basse

basse04 commented:
Hi,
done changes in README and thing-types.xml.
Missed to do mvn spotless:apply before I sent the files, could this be the reason for all failed checks?

I had to upgrade my IDE, due to missing dependencies, and after that is it not possible to install local bindings ( in my case ferroamp ) in UI.
I do as I always had done but after Settings / Things / + / Install Bindings / Am I directed to Community Marketplace ?
I have run the binding in another machine with OH 4.1.2-1 and it works to just put the *.jar in /usr/share/openhab/addons and the binding will show up in karaf / openhab:
I have found a solution at: https://community.openhab.org/t/building-openhab-with-the-latest-version-4-1-x/148853/4 but it doesn't work.
Do you maybe have any idea's how to solve this.
BR Basse

lsiepel commented on bundles/org.openhab.binding.ferroamp/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,978 @@
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+	<channel-type id="ehub-wloadconsq-l1">
```
Many types can be combined as only the label differs. The Label can be set on the channel declaration.

For instance: ehub-wloadconsq-l1, ehub-wloadconsq-l2 and ehub-wloadconsq-l3 are all the same except for the label.

Create one combined channel type:
```
	<channel-type id="ehub-wloadconsq">
		<item-type>Number:Energy</item-type>
		<label>WLoadConsq</label>
		<category>Energy</category>
		<state readOnly="true" pattern="%.2f %unit%"/>
	</channel-type>
```

and use it in the thing structure as:

```
    <channel typeId="ehub-wloadconsq" id="ehubwloadconsql1"/>
        <label>WLoadConsq L1</label>
    </channel>
    <channel typeId="ehub-wloadconsq" id="ehubwloadconsql2"/>
        <label>WLoadConsq L2</label>
    </channel>
    <channel typeId="ehub-wloadconsq" id="ehubwloadconsql3"/>
        <label>WLoadConsq L3</label>
    </channel>
```

This would reduce the amount of types dramatically

jlaur replied:
See also https://github.com/openhab/openhab-addons/pull/16258#discussion_r1615144427 regarding system channel types.

basse04 replied:
Hi,
done changes in channels.xml, regarded to reducing types.
I had to do it as in channel.xml otherwise mine IDE was complaining, hope this is OK.

basse04 replied:
Please see commit at 20240607

lsiepel commented on bundles/org.openhab.binding.ferroamp/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,34 @@
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<!-- ferroamp Energyhub Thing Type -->
+	<thing-type id="energyhub">
```
Not fixed yet.

basse04 replied:
Hi,
This is fixed now, 20240603.
About Ferroamp.xml so has there been some kind  of mismatch, only the file as ferroamp.xml should be in the /src/main/resources/thing

basse04 replied:
Same mismatch, Ferramp should not be, ferroamp instead

lsiepel replied:
Sorry i don;t understand what you mean.
I see two files that declare the same thing: thing-type.xml and ferroamp.xml and they have a conflicting label for instance.

basse04 replied:
Please, see commit at 20240607

basse04 replied:
Please, see commit 20240607

lsiepel commented on bundles/org.openhab.binding.ferroamp/src/main/java/org/openhab/binding/ferroamp/internal/FerroampBindingConstants.java:
```diff
@@ -0,0 +1,209 @@
+    public static final ThingTypeUID THING_TYPE_ENERGYHUB = new ThingTypeUID(BINDING_ID, "energyhub");
+
+    // List of EHUB Channel ids
+    public static final String CHANNEL_EHUBWLOADCONSQL1 = "ehubWLoadConsqL1";
```
Not fixed yet.

basse04 replied:
Fixed

lsiepel commented:
> Hi, done changes in README and thing-types.xml. Missed to do mvn spotless:apply before I sent the files, could this be the reason for all failed checks?
> 
> I had to upgrade my IDE, due to missing dependencies, and after that is it not possible to install local bindings ( in my case ferroamp ) in UI. I do as I always had done but after Settings / Things / + / Install Bindings / Am I directed to Community Marketplace ? I have run the binding in another machine with OH 4.1.2-1 and it works to just put the *.jar in /usr/share/openhab/addons and the binding will show up in karaf / openhab: I have found a solution at: https://community.openhab.org/t/building-openhab-with-the-latest-version-4-1-x/148853/4 but it doesn't work. Do you maybe have any idea's how to solve this. BR Basse

1. You can always check why it fails by clicking the details button next to the failed task. From that detaisl page look for some build output scroll and see the error. 
1. This can certainly be spotless. You can be sure to run spotless. Once completed you should get a summary and git would have detected file changes that can be comitted and pushed (deepends on your IDE on how that is visible)
1. I don't see how updating your IDE dependency's would have effect on a redirectino to community marketplace, sorry i just don't understand.

basse04 commented:
Hi @lsiepel, @jlaur 
Just sent some changes in, README, thing-types.xml, addons.xml
This time will not DCO check work as it has the very likely Commit Message , Author, Commiter as always, hm..
BR Basse

lsiepel commented:
> Hi @lsiepel, @jlaur Just sent some changes in, README, thing-types.xml, addons.xml This time will not DCO check work as it has the very likely Commit Message , Author, Commiter as always, hm.. BR Basse

Can you add a response to the comments you fixed. Each run i have to look for all changes to discover only a part are fixed.

basse04 commented:
Hi,
absolutly, I will do like that.
I will be in Kiel , Germany this weekend so next week.(I'm living in Sweden btw)
BR Basse

basse04 commented on bundles/org.openhab.binding.ferroamp/src/main/java/org/openhab/binding/ferroamp/internal/FerroampBindingConstants.java:
```diff
@@ -0,0 +1,209 @@
+    public static final ThingTypeUID THING_TYPE_ENERGYHUB = new ThingTypeUID(BINDING_ID, "energyhub");
+
+    // List of EHUB Channel ids
+    public static final String CHANNEL_EHUBWLOADCONSQL1 = "ehubwloadconsql1";
```
Hi,
I think this has been solved at some other change, but I can't be sure. Please let me know otherwise.

lsiepel replied:
Not yet fixed.

lsiepel commented on bundles/org.openhab.binding.ferroamp/README.md:
````diff
@@ -0,0 +1,222 @@
+## Items
+
+```java
+Number:Energy ferroamp "ehub-wloadconsq-l1" <energy> { channel="ferroamp:energyhub:myenergyhub:ehub-wloadconsq-l1" }
+String ferroamp "request-charge" <energy> { channel="ferroamp:energyhub:myenergyhub:request-charge" }
````
These items have the same name, that is not going to work.

basse04 replied:
Done changes 20240607

lsiepel replied:
Not fixed.

lsiepel commented on bundles/org.openhab.binding.ferroamp/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,1030 @@
+		<state readOnly="true" pattern="%.2f %unit%"/>
+	</channel-type>
+	<channel-group-type id="ehub-wloadconsq">
+		<label>EHUB, WLoadConsq</label>
```
```suggestion
		<label>WLoadConsq</label>
```

The thing structure has already improved much, thanks. When looking at these channels i see some more room for improvement. 
- all channels labels have a prefix that corrosponds to the thing it belongs to, making the prefix redundant. Please remove
- the term it self corrosponds to the api field name making it a technical name. I prefer to change this to a label that indicates the functional use
- there are still many channel types that are identical except for their label, these can be re-used.

hope this makes sense as this applies to all channels in this file.

basse04 replied:
Hi,
done changes in channels.xml and ferroamp.xml, Please have a look!
This time it even works, it just not look OK :-)

We have had conversion with Ferroamp about the information in the External API specification but this is not a high level issue for them so this is the closest we are. I liked to have better Labels but I don't know more than this, Sorry for that.

lsiepel commented on bundles/org.openhab.binding.ferroamp/src/main/java/org/openhab/binding/ferroamp/internal/FerroampHandler.java:
```diff
@@ -0,0 +1,778 @@
+        channelConfigEso = FerroampChannelConfiguration.getChannelConfigurationEso();
+        channelConfigEsm = FerroampChannelConfiguration.getChannelConfigurationEsm();
+
+        ferroampConfig = getConfigAs(FerroampConfiguration.class);
```
Don;t you need to check the configuration for sanity and set the thingstatus to CONFIGURATION_ERROR accordingly

basse04 replied:
Ongoing 20240607

lsiepel commented on bundles/org.openhab.binding.ferroamp/src/main/java/org/openhab/binding/ferroamp/internal/FerroampHandler.java:
```diff
@@ -0,0 +1,778 @@
+            boolean thingReachable = true;
+
+            if (thingReachable) {
+                updateStatus(ThingStatus.ONLINE);
+                refresh();
```
This would possibly flip the Thing from unknown to online to offline.
I recommend to extract the `get/sendMQTT` method to a seperate class and throw custom exceptions if something fails.
This would make that seperate class responsible for communication with the remote mqtt service and would make this ThingHandler responsible to update the channels / thing status.

basse04 replied:
Ongoing 20240607

lsiepel reviewed (changes requested):
Left some new comments, also some comments from the previous review are not fixed or commented on yet. You might have missed them as they are behind the 'show more' button.

basse04 commented on bundles/org.openhab.binding.ferroamp/src/main/java/org/openhab/binding/ferroamp/internal/FerroampHandler.java:
```diff
@@ -0,0 +1,785 @@
+    private void processIncomingJsonMessage(String topic, String messageJson) {
+        if ("extapi/data/ehub".equals(topic)) {
+            String[] ehubChannelPostsValue = new String[86]; // Array for EHUB (Energy Hub) Posts
+            JsonObject jsonElementsObject = new Gson().fromJson(new Gson().fromJson(messageJson, JsonObject.class),
```
Ongoing 20240607

basse04 commented on bundles/org.openhab.binding.ferroamp/src/main/java/org/openhab/binding/ferroamp/internal/FerroampHandler.java:
```diff
@@ -0,0 +1,785 @@
+        localConfigurationConnection.start();
+        localConfigurationConnection.setCredentials(ferroampConfig.userName, ferroampConfig.password);
+
+        if (localConfigurationConnection != null) {
```
Ongoing 20240607

basse04 commented on bundles/org.openhab.binding.ferroamp/src/main/java/org/openhab/binding/ferroamp/internal/FerroampHandler.java:
```diff
@@ -0,0 +1,785 @@
+        String valueConfiguration = command.toString();
+
+        if (FerroampBindingConstants.CHANNEL_REQUESTCHARGE.equals(channelUID.getId())) {
+            String requestCmdJsonCharge = "{\"" + "transId" + "\":\"" + transId
+                    + "\",\"cmd\":{\"name\":\"charge\",\"arg\":\"" + valueConfiguration + "\"}}";
```
Ongoing 20240607

lsiepel replied:
Please ping me when you are ready. 10 days left for the next stable release.

basse04 commented:
Hi, done changes 20240607. 
Adviced changes done in channels.xml, ferroamp.xml, README.md,  Please, have a look.
Changes done in EhubJsonElements.java, EhubParameters.java, EsoParameters.java, FerroampBindingConstants.java, FerroampChannelConfiguration.java, FerroampHandler.java due to found faults in code.

At this Commit won't DCO work, I use the very same Author and Committer every time, some times it works others not ?
Örjan Backsell <orjan.backsell@gmail.com>

Br Basse

lsiepel commented:
> Hi, done changes 20240607. Adviced changes done in channels.xml, ferroamp.xml, README.md, Please, have a look. Changes done in EhubJsonElements.java, EhubParameters.java, EsoParameters.java, FerroampBindingConstants.java, FerroampChannelConfiguration.java, FerroampHandler.java due to found faults in code.
> 
> At this Commit won't DCO work, I use the very same Author and Committer every time, some times it works others not ? Örjan Backsell [orjan.backsell@gmail.com](mailto:orjan.backsell@gmail.com)
> 
> Br Basse

Some small things left, see open comments. I see one larger issue that allready had somecomments, but has not improved much yet. This binding has the channel label and the json property name tied together as one string. They should be uncoupled.
The json property name (ferroamp domain) is not as meaningfull / user friendly to present to the user (channel label, openHAB domain) .

basse04 commented:
Hi,
First of all, many Thank's for all advices.
I will have a look asap, but I will be on vacation for some weeks first, with limited time for this.
I have already done some changes, but they are still under testing and not uploaded yet.
I still have the issue with DTO to solve, I have an solution regarded to start/stop and some other things that you have advice me to change.
I will let you know when these are ready for you to check.
BR Basse

wborn commented on bundles/org.openhab.binding.ferroamp/pom.xml:
```diff
@@ -0,0 +1,26 @@
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>4.2.0-SNAPSHOT</version>
```
```suggestion
    <version>4.3.0-SNAPSHOT</version>
```

lsiepel commented:
Could you update your main branch from openhab/main and then update your basse04:ferroamp_v3 branch from your basse04:main (should be a few clicks on github) that would make it a lot easier for me to review and build.

lsiepel commented:
Thanks for updating that quick. I'm getting a better understanding of this binding now. 
Unfortunately i see 100+ compile warnings and they need to be addressed, but before you jump in i like to discuss some parts.

I found this file: https://ferroamp.com/wp-content/uploads/2023/01/Ferroamp-External-API-specification-E.pdf and that shows some example JSON messages. I assume you use the same JSON messages in this binding. 

This is relevant because i think the binding would be better maintainable if a dto folder is created with the java classes mirroring this JSON schema. It would make the `FerroampMqttCommunication` much simpler as GSON can be used to directly serialize to the DTO classes. It would also prevent many compile warnings.

Other topic would be the channel names. In the past reviews i left some comments, it was adapted and improved. I would propose the following changes:
1. Strip the Thing abbrevation from the channel and/or channel type (e.g. `ehubwloadconsql1` => `wloadconsql1`)
2. Use the naming convention for both channel and type id's (e.g. `w-load-cons-q-l1`)
3. Use more meaningfull channel id's where possible. (e.g. `load-consumed-l1`)
4. Use more meaningfull channel labels where possible. (e.g. `Load Energy Consumed L1`)
5. Re-use channel type's 

Agree, the documentation is not clear on what the value actually represents, but with the app and the internet, you can probably find allmost all. A good source could be HA: https://github.com/henricm/ha-ferroamp/blob/master/custom_components/ferroamp/sensor.py

While the code has improved, i think these are the last two mayor issues to fix. Let me know if i can be of any help to progress.

basse04 commented:
Hi,
you are right, I have used this document for JSON-messages.
I will have a look at your latest advices and also start the work for to have an DTO class implemented asap, I will let you know.
I will ask for help if or rather when ;-) I need it, be so sure.
Br Basse

basse04 commented:
Hi,
just a small report:
I have almost finished the 5 steps you adviced me to ( some work, but as I now see it very necessary ). I will now start with the GSON and DTO issues.
I will let you know asap.
I wish you a nice weekend!
Br Basse

lsiepel commented:
> Hi, just a small report: I have almost finished the 5 steps you adviced me to ( some work, but as I now see it very necessary ). I will now start with the GSON and DTO issues. I will let you know asap. I wish you a nice weekend! Br Basse

Take your time, no need to rush.

basse04 commented:
Hi,
Thank's a lot for your patience. I have now done changes to the channel-names etc., as you adviced me to. And as I said before,it looks much better now.
I have also done a lot of other changes and add new classes, so this is almost like a new binding. There are things left to do, but I think if you will have a look, it will be very good, and  if I then can leave some things that hopefully will be OK and work on with others.
I have the binding working now, but there has been problems, regarded to parse the different messages. I have tried to insted of how it is now regarded to ex. EHUB part, have a loop with the very same object for every new post and just put every new value to String[] ehubChannelPostsValue and then update when all values are fulfilled, but I have not figure out how to have it work like that. I will continue with that, and every tip is as ususal very helpful.
BR Basse

lsiepel commented:
> Hi, Thank's a lot for your patience. I have now done changes to the channel-names etc., as you adviced me to. And as I said before,it looks much better now. I have also done a lot of other changes and add new classes, so this is almost like a new binding. There are things left to do, but I think if you will have a look, it will be very good, and if I then can leave some things that hopefully will be OK and work on with others. I have the binding working now, but there has been problems, regarded to parse the different messages. I have tried to insted of how it is now regarded to ex. EHUB part, have a loop with the very same object for every new post and just put every new value to String[] ehubChannelPostsValue and then update when all values are fulfilled, but I have not figure out how to have it work like that. I will continue with that, and every tip is as ususal very helpful. BR Basse

Di you push your changes? as the last commit is from Aug 20, 2024 and i don;t see any fixes checked in.
There is also a merge conflict, maybe thet prevented the push as it probably needs some interactions.

basse04 commented:
Hi,
you are absolutly rtight, there are now changes. So  now I have to figure out why not. Due to Eclipse so is it no Unstaged Changes left, strange. I let you know.
Basse

basse04 commented:
Hi, 
I can now see the changes in https://github.com/basse04/openhab-addons/tree/ferroamp_v3/bundles, but somehow has the PR been closed ?
Basse

lsiepel commented:
Can't you re-open this PR on the bottom of this page? Otherwise you might create a new PR that is also fine.

basse04 commented:
I have tried with the Reopen...., but this one is grayed out. and not working due to that " There are no new commits...."
I have had similar problems before when Commiting from Eclipse. I am quit sue that there is something I do wrong or miss to do but I can't find what, some times everything works as expected others not.
Eclipse take away all files from Staging at commit, but some how will they not be sent to the repisitory. About Closed is it even more confusing, I have not closed anything.
But actually no problem, I will make a new PR, but first will I update the README that I totally have forgot about.

lsiepel commented:
> I have tried with the Reopen...., but this one is grayed out. and not working due to that " There are no new commits...." I have had similar problems before when Commiting from Eclipse. I am quit sue that there is something I do wrong or miss to do but I can't find what, some times everything works as expected others not. Eclipse take away all files from Staging at commit, but some how will they not be sent to the repisitory. About Closed is it even more confusing, I have not closed anything. But actually no problem, I will make a new PR, but first will I update the README that I totally have forgot about.

Usually when you commit, you commit the changes to your _local_ fork. To you need to explicitly push them to the remote fork to end up here. But as you said your branch at GH is uptodate it is kind of weird. I would suggest to just try to create a new PR.
